### PR TITLE
Revise SkyDB API

### PR DESCRIPTION
**Pull request by mrcnski**

### CODE REVIEW

File: src/crypto.ts [hunk 1]
```diff
@@ -17,19 +21,9 @@ export function HashAll(...args: any[]): Uint8Array {
...
-
-// HashFileID hashes the given fileID
-export function HashFileID(fileID: FileID): Uint8Array {
-  return HashAll(
-    encodeNumber(fileID.version),
-    encodeString(fileID.applicationID),
-    encodeNumber(fileID.fileType),
-    encodeString(fileID.filename)
-  );
+// Ashes the given registry entry
```
Comment (peterjan) [inline src/crypto.ts]: update docstring

File: src/skydb.ts [hunk 2]
```diff
@@ -1,105 +1,65 @@
...
+  const entry = await this.registry.lookup(publicKey, dataKey);
+  if (entry === null) {
     throw new Error("not found");
   }
 
-  // download the data in that Skylink
-  const skylink = existing.value.data;
+  // Download the data in that Skylink
+  // TODO: Replace with download request method.
+  const skylink = entry.value.data;
```
Comment (peterjan) [inline src/skydb.ts]: should we validate this to be a skylink and throw a meaningful error?

File: src/skydb.ts [hunk 3]
```diff
@@ -1,105 +1,65 @@
...
-  if (!existing) {
+  const entry = await this.registry.lookup(publicKey, dataKey);
+  if (entry === null) {
     throw new Error("not found");
   }
 
-  // download the data in that Skylink
-  const skylink = existing.value.data;
+  // Download the data in that Skylink
+  // TODO: Replace with download request method.
```
Comment (peterjan) [inline src/skydb.ts:32]: remove TODO?

File: src/skydb.ts [hunk 4]
```diff
@@ -1,105 +1,65 @@
...
-
-  // fetch the current value to find out the revision, timeout after 2s
-  let existing: SignedRegistryValue | null;
-  try {
-    existing = await promiseTimeout(this.lookupRegistry(user, fileID), 2000);
-  } catch (error) {
-    existing = null;
+export async function setJSON(this: SkynetClient, privateKey: SecretKey, dataKey: string, json: object, revision: number = -1): Promise<boolean> {
+  // Upload the data to acquire its skylink
+  // TODO: Replace with upload request method.
```
Comment (peterjan) [inline src/skydb.ts:58]: remove TODO

File: src/skydb.ts [hunk 5]
```diff
@@ -1,105 +1,65 @@
...
+  const response = await this.executeRequest({
+    ...this.customOptions,
+    method: "post",
+    endpointPath: "/skynet/skyfile",
+    data: json.toString(),
+  });
+  const skylink = response.skylink;
+
+  const publicKey = pki.ed25519.publicKeyFromPrivateKey({privateKey});
+  if (revision === -1) {
```
Comment (peterjan) [inline src/skydb.ts]: did we land on this behavior? (works for me)

Comment (peterjan) [reply src/skydb.ts]: the reason I ask is because I guess the `-1` convention might be considered somewhat of a code smell, even though I don't mind that type of thing

File: src/crypto.ts [hunk 6]
```diff
@@ -1,7 +1,11 @@
 import blake from "blakejs";
-import { RegistryValue } from "./registry";
-import { FileID } from "./skydb";
+import { RegistryEntry } from "./registry";
 import { stringToUint8Array } from "./utils";
+import { pki } from "node-forge";
+
+export type PublicKey = pki.ed25519.NativeBuffer;
```
Comment (peterjan) [inline src/crypto.ts:6]: does this make sense to export? if so, shouldn't we re-export these in index.ts?

File: src/skydb.ts [hunk 7]
```diff
@@ -1,105 +1,65 @@
...
+    // fetch the current value to find out the revision.
+    const entry = await this.registry.lookup(publicKey, dataKey)
+    if (entry) {
+      revision = entry.revision + 1;
+    } else {
+      revision = 0;
+    }
   }
 
   // TODO: we could (/should?) verify here
```
Comment (peterjan) [inline src/skydb.ts]: we should implement this TODO in the next version imo

File: src/skydb.ts (hunk 3 above)
Comment (mrcnski) [reply src/skydb.ts:32]: I can't yet, the request methods are not available yet, one is in an active PR

File: src/skydb.ts (hunk 5 above)
Comment (mrcnski) [reply src/skydb.ts]: Dunno, I got it from your proposal. I think it's fine as a default value but maybe it's idiosyncratic for JS.

File: src/registry.ts [hunk 8]
```diff
@@ -56,8 +43,7 @@ export async function lookupRegistry(
 
   if (response.status === 200) {
     return {
-      value: {
-        tweak: Uint8Array.from(Buffer.from(response.data.tweak)),
+      entry: {
         data: Buffer.from(hexToUint8Array(response.data.data)).toString(),
         revision: parseInt(response.data.revision, 10),
```
Comment (interstella-5555) [inline src/registry.ts:64]: Is `revision` a string in api?

File: src/registry.ts [hunk 9]
```diff
@@ -56,8 +43,7 @@ export async function lookupRegistry(
 
   if (response.status === 200) {
     return {
-      value: {
-        tweak: Uint8Array.from(Buffer.from(response.data.tweak)),
+      entry: {
```
Comment (interstella-5555) [inline src/registry.ts:60]: why do we need an `entry` wrapper? couldn't we just return

```javascript
return {
  data: Buffer.from(hexToUint8Array(response.data.data)).toString(),
  revision: parseInt(response.data.revision, 10),
  signature: response.data.signature,
}
```

File: src/registry.ts [hunk 10]
```diff
@@ -67,13 +53,7 @@ export async function lookupRegistry(
   throw new Error(`unexpected response status code ${response.status}`);
```
Comment (interstella-5555) [inline src/registry.ts]: You throw on some response codes and some response codes return null. Should we be consistent here or is there any particular reason for that?

File: src/registry.ts [hunk 11]
```diff
@@ -1,52 +1,39 @@
...
+  signature: Signature;
 };
 
-export async function lookupRegistry(
-  this: SkynetClient,
-  user: User,
-  fileID: FileID,
-  customOptions = {}
-): Promise<SignedRegistryValue | null> {
+export async function lookup(this: SkynetClient, publickey: PublicKey, datakey: string, customOptions = {}): Promise<SignedRegistryEntry|null> {
```
Comment (interstella-5555) [inline src/registry.ts]: Why is this called `lookup`? It's non standard naming for a getter - would call it a `getRegistryEntry` or `registry.getEntry` if we already namespace with `registry`.

File: src/registry.ts [hunk 12]
```diff
@@ -88,17 +68,12 @@ export async function updateRegistry(
...
-          applicationid: fileID.applicationID,
-          filetype: fileID.fileType,
-          filename: fileID.filename,
-        },
-        revision: srv.value.revision,
-        data: Array.from(Uint8Array.from(Buffer.from(srv.value.data))),
-        signature: Array.from(Uint8Array.from(srv.signature)),
+        datakey,
+        revision: entry.revision,
+        data: Array.from(Uint8Array.from(Buffer.from(entry.data))),
```
Comment (interstella-5555) [inline src/registry.ts]: Are you sure these 3 are needed? Wouldn't this suffice? Or even maybe just `Array.from` ? (not tested it, just asking because it looks pretty entangled)
```suggestion
        data: Array.from(Buffer.from(entry.data)),
```

File: src/skydb.ts [hunk 13]
```diff
@@ -1,105 +1,79 @@
...
-
-// getFile will lookup the entry for given skappID and filename, if it exists it
-// will try and download the file behind the skylink it has found in the entry.
-export async function getFile(this: SkynetClient, user: User, fileID: FileID): Promise<SkyFile> {
+export async function getJSON(this: SkynetClient, publicKey: PublicKey, dataKey: string): Promise<{ json: object, revision: number }|null> {
   // lookup the registry entry
-  const existing = await this.lookupRegistry(user, fileID);
-  if (!existing) {
+  const entry = await this.registry.lookup(publicKey, dataKey);
+  if (entry === null) {
```
Comment (interstella-5555) [inline src/skydb.ts:27]: I would just return `null` on not found, catching errors is just the worst

File: src/skydb.ts [hunk 14]
```diff
@@ -1,105 +1,79 @@
...
     throw new Error("not found");
   }
 
-  // download the data in that Skylink
-  const skylink = existing.value.data;
+  // Download the data in that Skylink
+  // TODO: Replace with download request method.
+  let skylink: string;
+  try {
+    skylink = parseSkylink(entry.value.data);
```
Comment (interstella-5555) [inline src/skydb.ts]: `parseSkylink` should probably return `null` on not found as well... not saying to change it here but just sayin'

File: src/skydb.ts [hunk 15]
```diff
@@ -1,105 +1,79 @@
...
+  } catch (error) {
+    throw new Error(`invalid skylink: ${error}`);
+  }
```
Comment (interstella-5555) [inline src/skydb.ts]: `parseSkylink` already throws a meaningful error, no need to catch it and rethrow simplified version of it

```suggestion
  const skylink = parseSkylink(entry.value.data);
```

File: src/skydb.ts [hunk 16]
```diff
@@ -1,105 +1,79 @@
...
     ...this.customOptions,
     method: "get",
     url: this.getSkylinkUrl(skylink),
   });
 
-  // wrap the data in a skyfile and return it
-  const metadata = JSON.parse(response.headers["skynet-file-metadata"]);
-  const file = new SkyFile(new File([response.data], metadata.filename, { type: response.headers["content-type"] }));
-  return file;
+  return { json: response.data, revision: entry.revision };
```
Comment (interstella-5555) [inline src/skydb.ts]: Let's just call it `data` so it's more generic and we could use the same response type in future `getText` or `getBuffer` methods. We already know it's json because it's the response from `getJson`. Actually, it's not even json - it's json parsed as an object :)

```suggestion
  return { data: response.data, revision: entry.revision };
```

File: src/skydb.ts [hunk 17]
```diff
@@ -1,105 +1,79 @@
...
+export async function setJSON(this: SkynetClient, privateKey: SecretKey, dataKey: string, json: object, revision: number = -1): Promise<boolean> {
+  // Upload the data to acquire its skylink
+  // TODO: Replace with upload request method.
+  const response = await this.executeRequest({
+    ...this.customOptions,
+    method: "post",
+    endpointPath: "/skynet/skyfile",
+    data: json.toString(),
+  });
+  const skylink = response.skylink;
```
Comment (interstella-5555) [inline src/skydb.ts]:
```suggestion
  const { skylink } = await this.executeRequest({
    ...this.customOptions,
    method: "post",
    endpointPath: "/skynet/skyfile",
    data: json.toString(),
  });
```

File: src/skydb.ts [hunk 18]
```diff
@@ -1,105 +1,79 @@
...
+
+  const publicKey = pki.ed25519.publicKeyFromPrivateKey({privateKey});
+  if (revision === -1) {
```
Comment (interstella-5555) [inline src/skydb.ts]: I would probably make `revision` an optional argument and check `if (revision) {` here instead.

File: src/skydb.ts [hunk 19]
```diff
@@ -1,105 +1,79 @@
...
-  const signature = user.sign({ message: HashRegistryValue(value) });
+  const signature = pki.ed25519.sign({
+    message: HashRegistryEntry(entry),
+    privateKey,
+  });
 
   // update the registry
-  const updated = await this.updateRegistry(user, fileID, { value, signature });
+  const updated = await this.registry.update(publicKey, dataKey, entry, signature);
   return updated;
```
Comment (interstella-5555) [inline src/skydb.ts]: No need to await returned value.

```suggestion
  return this.registry.update(publicKey, dataKey, entry, signature);
```

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/registry.ts (hunk 10 above)
Comment (mrcnski) [reply src/registry.ts]: I don't know, cc @peterjanbrone

Comment (mrcnski) [reply src/registry.ts]: @kwypchlo what is more idiomatic, throwing or returning null?

File: src/registry.ts (hunk 8 above)
Comment (mrcnski) [reply src/registry.ts:64]: It's a number

File: src/registry.ts (hunk 9 above)
Comment (mrcnski) [reply src/registry.ts:60]: I guess to emphasize that the `signature` is on the entire `entry` object? @peterjanbrone

File: src/registry.ts (hunk 11 above)
Comment (mrcnski) [reply src/registry.ts]: I like `getEntry` but this will require another discussion on Discord, hopefully a short one, but we should get everyone's input.

File: src/skydb.ts (hunk 14 above)
Comment (mrcnski) [reply src/skydb.ts]: Wouldn't that be a breaking change? I can create a followup issue.

File: src/skydb.ts (hunk 17 above)
Comment (mrcnski) [reply src/skydb.ts]: Looks like you commented on an outdated commit

File: src/registry.ts (hunk 10 above)
Comment (peterjan) [reply src/registry.ts]: Yeah I decided to throw here and return null because this is basically a sanity check to catch developer error.
That or the siad API changed, we basically do not expect to throw this error ever and I wanted to avoid the `try-catch` on the caller side if we want to stay consistent in the expected case.

It's kind of like `build.Critical` or `panic` in siad, at least that was my intention.

File: src/registry.ts (hunk 9 above)
Comment (peterjan) [reply src/registry.ts:60]: Yeah that's the reason.

File: src/registry.ts (hunk 8 above)
Comment (peterjan) [reply src/registry.ts:64]: it's a uint64 right so technically we need to handle that, perhaps we should leave a TODO or comment at least?

File: src/registry.ts (hunk 10 above)
Comment (interstella-5555) [reply src/registry.ts]: @peterjanbrone @m-cat SDK should probably abstract that away, otherwise we will end up with people having to check if the response is not null AND also writing a try-catch.

File: src/skydb.ts [hunk 20]
```diff
@@ -37,13 +37,8 @@ export async function setJSON(
...
   // Upload the data to acquire its skylink
   // TODO: Replace with upload request method.
-  const { data } = await this.executeRequest({
-    ...this.customOptions,
-    method: "post",
-    endpointPath: "/skynet/skyfile",
-    data: json.toString(),
-  });
-  const skylink = data.skylink;
+  const file = new File([JSON.stringify(json)], dataKey, { type: "text/plain" });
```
Comment (interstella-5555) [inline src/skydb.ts]:
```suggestion
  const file = new File([JSON.stringify(json)], dataKey, { type: "application/json" });
```

File: src/registry.ts [hunk 21]
```diff
@@ -97,5 +97,5 @@ export async function setEntry(
   if (response.status === 204) {
     return true;
   }
-  throw new Error(`unexpected response status code ${response.status}`);
+  return false;
```
Comment (interstella-5555) [inline src/registry.ts]: `setEntry` should probably be just. In other case it fails, let's say because revision was wrong, user would not get any error message and would not know how to act - either retry or change the payload, returning `false` doesn't tell user anything useful.

```javascript
return this.executeRequest({
           ...opts,
      	      method: "post",
      	      data: {
      	        publickey: {
                          algorithm: "ed25519",
          	          key: Array.from(publickey),
       	        },
        	        datakey,
        	        revision: entry.revision,
        	        data: Array.from(Buffer.from(entry.data)),
        	        signature: Array.from(signature),
	      },
});
```

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/integration.test.ts [hunk 22]
```diff
@@ -0,0 +1,31 @@
+import { SkynetClient } from "./client";
```
Comment (interstella-5555) [inline src/integration.test.ts:1]: let's move this integration test out of `src` to `e2e` directory

File: src/crypto.ts [hunk 23]
```diff
@@ -17,19 +21,9 @@ export function HashAll(...args: any[]): Uint8Array {
...
-// HashFileID hashes the given fileID
-export function HashFileID(fileID: FileID): Uint8Array {
-  return HashAll(
-    encodeNumber(fileID.version),
-    encodeString(fileID.applicationID),
-    encodeNumber(fileID.fileType),
-    encodeString(fileID.filename)
-  );
+// Hashes the given registry entry.
+export function HashRegistryEntry(registryEntry: RegistryEntry): Uint8Array {
```
Comment (peterjan) [inline src/crypto.ts:30]: this is missing the datakey, you will run into signature verification issues if you do not include it

File: src/integration.test.ts [hunk 24]
```diff
@@ -6,18 +6,17 @@ import { pki } from "node-forge";
 const { publicKey, privateKey } = pki.ed25519.generateKeyPair();
 const client = new SkynetClient("https://siasky.dev");
 
-const appID = "HelloWorld";
+const datakey = "HelloWorld";
 const json = { data: "thisistext" };
 
 // skip - used to verify end-to-end flow
-describe.skip("siasky.dev end to end", () => {
+describe("siasky.dev end to end", () => {
```
Comment (interstella-5555) [inline src/integration.test.ts]: skip that test please, it should be run only manually, otherwise there is a good chance it will fail on our CI and it's not a unit test

File: src/skydb.ts [hunk 25]
```diff
@@ -88,5 +89,5 @@ export async function setJSON(
 
   console.log(9);
   // update the registry
-  return this.registry.setEntry(publicKey, dataKey, entry, signature);
+  this.registry.setEntry(publicKey, dataKey, entry, signature);
```
Comment (interstella-5555) [inline src/skydb.ts]: if you don't return here, you will need to add `await`

File: src/skydb.ts [hunk 26]
```diff
@@ -49,7 +49,7 @@ export async function setJSON(
   if (!revision) {
     // fetch the current value to find out the revision.
     console.log(3);
-    const entry = await this.registry.getEntry(publicKey, dataKey);
+    const entry = await promiseTimeout(this.registry.getEntry(publicKey, dataKey), 5000);
```
Comment (interstella-5555) [inline src/skydb.ts]: we shouldn't add an arbitrary timeout here, we should rely on siad either 404 or 200

**Review by interstella-5555 — CHANGES_REQUESTED** [review]
please remove all `console.log` statements left from debugging

File: src/integration.test.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import { SkynetClient } from "./client";
+// import { Buffer } from "buffer";
+import { pki } from "node-forge";
+// import { readData } from "./utils";
```
Comment (peterjan) [inline src/integration.test.ts]: remove unused imports

File: src/integration.test.ts [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+
+const { publicKey, privateKey } = pki.ed25519.generateKeyPair();
+const client = new SkynetClient("https://siasky.dev");
+
+const datakey = "HelloWorld";
+const json = { data: "thisistext" };
+
+// skip - used to verify end-to-end flow
+describe.skip("siasky.dev end to end", () => {
+  it("to update the file in the SkyDB", async () => {
```
Comment (peterjan) [inline src/integration.test.ts]: could improve on these descriptions

`describe.skip("skydb integration test")`
`it("should be able to both `setJSON` and `getJSON`")`

File: src/integration.test.ts [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+    // set the file in the SkyDB
+    await client.db.setJSON(privateKey, datakey, json);
+
+    // get the file in the SkyDB
+    const actual = await client.db.getJSON(publicKey, datakey);
```
Comment (peterjan) [inline src/integration.test.ts:17]: Should we be returning the revision number here? I'm not too sure, if the caller cares about it he can just use the registry himself and not go through `setJSON` and `getJSON` which essentially are convenience methods right.

I think `setJSON` and `getJSON` should mirror each other as much as possible and thus only return json.

File: src/registry.ts [hunk 30]
```diff
@@ -1,52 +1,45 @@
...
   customOptions = {}
-): Promise<SignedRegistryValue | null> {
+): Promise<SignedRegistryEntry | null> {
   const opts = {
     ...defaultRegistryOptions,
     ...this.customOptions,
     ...customOptions,
   };
 
+  const userID = publickey.toString("hex");
```
Comment (peterjan) [inline src/registry.ts]: (nit) would not call this `userID` and simply inline:
`publickey: ed25519:${publickey.toString("hex")}`

File: src/registry.ts [hunk 31]
```diff
@@ -56,58 +49,47 @@ export async function lookupRegistry(
...
+    publickey: {
+      algorithm: "ed25519",
+      key: Array.from(publickey),
+    },
+    datakey: Buffer.from(HashDataKey(datakey)).toString("hex"),
+    revision: entry.revision,
+    data: Array.from(Buffer.from(entry.data)),
+    signature: Array.from(signature),
+  };
+  console.log(data);
```
Comment (peterjan) [inline src/registry.ts]: remove debug code

File: src/skydb.test.ts [hunk 32]
```diff
@@ -1,42 +1,23 @@
...
+const json = { data: "thisistext" };
 
 const portalUrl = defaultSkynetPortalUrl;
 const registryUrl = `${portalUrl}/skynet/registry`;
 const uploadUrl = `${portalUrl}/skynet/skyfile`;
 
 const client = new SkynetClient(portalUrl);
 
-describe.skip("getFile", () => {
+describe("getFile", () => {
```
Comment (peterjan) [inline src/skydb.test.ts]: we do not have `getFile` and `setFile` anymore

File: src/skydb.test.ts [hunk 33]
```diff
@@ -47,19 +28,11 @@ describe.skip("getFile", () => {
   it("should perform a lookup and update the window to the skylink url", async () => {
```
Comment (peterjan) [inline src/skydb.test.ts]: outdated descriptions

File: src/skydb.test.ts [hunk 34]
```diff
@@ -47,19 +28,11 @@ describe.skip("getFile", () => {
...
-      publickey: `ed25519:${user.id}`,
-      fileid: Buffer.from(
-        JSON.stringify({
-          version: fileID.version,
-          applicationid: fileID.applicationID,
-          filetype: fileID.fileType,
-          filename: fileID.filename,
-        })
-      ).toString("hex"),
+      publickey: `ed25519:${publicKey}`,
```
Comment (peterjan) [inline src/skydb.test.ts]: this is confusing I think, in the actual code we are hex encoding this right?

Comment (peterjan) [reply src/skydb.test.ts]: maybe we need a helper function for this that also prepends the algorithm

File: src/skydb.ts [hunk 35]
```diff
@@ -1,105 +1,93 @@
...
+export async function getJSON(
+  this: SkynetClient,
+  publicKey: PublicKey,
+  dataKey: string
+): Promise<{ data: Record<string, unknown>; revision: number } | null> {
   // lookup the registry entry
-  const existing = await this.lookupRegistry(user, fileID);
-  if (!existing) {
-    throw new Error("not found");
+  console.log(10);
```
Comment (peterjan) [inline src/skydb.ts]: remove debug code (commenting once but there's multiple in here)

File: src/skydb.test.ts [hunk 36]
```diff
@@ -1,79 +1,54 @@
...
+describe("getJSON", () => {
   let mock: MockAdapter;
 
   beforeEach(() => {
     mock = new MockAdapter(axios);
     mock.resetHistory();
   });
 
-  it("should perform a lookup and update the window to the skylink url", async () => {
+  // TODO
```
Comment (peterjan) [inline src/skydb.test.ts:27]: leftover TODO

File: src/client.ts [hunk 37]
```diff
@@ -62,12 +62,16 @@ export class SkynetClient {
   resolveHns = resolveHns;
 
   // SkyDB
-  getFile = getFile;
-  setFile = setFile;
+  db = {
```
Comment (peterjan) [inline src/client.ts:65]: not a priority now I guess but this will not be typed properly in the SkynetClient I don't think

**Review by peterjan — APPROVED** [review]

File: src/skydb.test.ts (hunk 36 above)
Comment (mrcnski) [reply src/skydb.test.ts:27]: I set the test to `skip` because it hasn't been fully written yet.

File: src/skydb.ts [hunk 38]
```diff
@@ -1,105 +1,81 @@
...
+    if (entry) {
+      // verify here
+      if (
+        !pki.ed25519.verify({
+          message: HashRegistryEntry(entry.entry),
+          signature: entry.signature,
+          publicKey,
+        })
+      ) {
+        throw new Error("could not verify signature");
```
Comment (interstella-5555) [inline src/skydb.ts]: Can we make this error a little more user friendly? Give some more details on what could be wrong.

File: src/client.ts (hunk 37 above)
Comment (peterjan) [reply src/client.ts:65]: easily fixed by the way, just define an interface 

```
export interface IDB {
  getJSON(
    publicKey: PublicKey,
    dataKey: string
  ): Promise<Record<string, unknown> | null>
  
  setJSON(
    privateKey: SecretKey,
    dataKey: string,
    json: Record<string, unknown>,
    revision?: number
  ): void;
}
```
and do 
```
  resolveHns = resolveHns;

  // SkyDB
  db: IDB = {
    getJSON: getJSON.bind(this),
  ...
```

(rename the interface, it's just for demonstrating purposes)

File: src/skydb.ts [hunk 39]
```diff
@@ -26,9 +24,8 @@ export async function getJSON(
     method: "get",
     url: this.getSkylinkUrl(skylink),
   });
-  console.log(12);
 
-  return { data: response.data, revision: entry.entry.revision };
+  return response.data;
```
Comment (interstella-5555) [inline src/skydb.ts]: Will we still have access to revision ?

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/skydb.ts (hunk 39 above)
Comment (mrcnski) [reply src/skydb.ts]: Yeah, PJ made the point that the registry functions can be called manually if you need the revision. The SkyDB functions should be easy to use without the user having to keep track of the revision. Makes sense to me IMO.

File: src/client.ts (hunk 37 above)
Comment (mrcnski) [reply src/client.ts:65]: Not so easily-fixed, I get typescript errors when I do this. I'm just going to make a follow-up issue for later, thanks.

File: src/skydb.ts (hunk 39 above)
Comment (interstella-5555) [reply src/skydb.ts]: Not really, let's say I want to update the content that I just fetched, I will need the revision for that.

File: src/skydb.ts (hunk 14 above)
Comment (mrcnski) [reply src/skydb.ts]: https://github.com/NebulousLabs/skynet-js/issues/143

File: src/client.ts (hunk 37 above)
Comment (mrcnski) [reply src/client.ts:65]: https://github.com/NebulousLabs/skynet-js/issues/144

File: src/skydb.ts (hunk 39 above)
Comment (mrcnski) [reply src/skydb.ts]: You would either use the registry from the beginning or just `setJSON`.

File: src/skydb.ts [hunk 40]
```diff
@@ -1,105 +1,82 @@
...
-export async function getFile(this: SkynetClient, user: User, fileID: FileID): Promise<SkyFile> {
+import { HashRegistryEntry, PublicKey, SecretKey } from "./crypto";
+import { RegistryEntry, SignedRegistryEntry } from "./registry";
+import { parseSkylink, promiseTimeout, trimUriPrefix, uriSkynetPrefix } from "./utils";
+
+export async function getJSON(
+  this: SkynetClient,
+  publicKey: PublicKey,
+  dataKey: string
+): Promise<{ data: Record<string, unknown>; revision: number } | null> {
```
Comment (peterjan) [inline src/skydb.ts:19]: (nit call it `json`?)

File: src/skydb.ts [hunk 41]
```diff
@@ -1,105 +1,82 @@
...
+  // TODO: Replace with upload request method.
+  const file = new File([JSON.stringify(json)], dataKey, { type: "application/json" });
+  const { skylink } = await this.uploadFileRequest(file, this.customOptions);
+
+  const publicKey = pki.ed25519.publicKeyFromPrivateKey({ privateKey });
+  if (!revision) {
+    // fetch the current value to find out the revision.
+    let entry: SignedRegistryEntry;
+    try {
+      entry = await promiseTimeout(this.registry.getEntry(publicKey, dataKey), 5000);
```
Comment (peterjan) [inline src/skydb.ts]: I think we should go with 30s

File: src/skydb.ts [hunk 42]
```diff
@@ -25,7 +25,7 @@ export async function getJSON(
     url: this.getSkylinkUrl(skylink),
   });
 
-  return { data: response.data, revision: entry.entry.revision };
+  return { json: response.data, revision: entry.entry.revision };
```
Comment (interstella-5555) [inline src/skydb.ts]: It should be `data` so once we have `getText` it will have the same interface instead of `entry.text`. Why did you change it in the first place?

File: src/skydb.ts [hunk 43]
```diff
@@ -45,7 +45,7 @@ export async function setJSON(
     // fetch the current value to find out the revision.
     let entry: SignedRegistryEntry;
     try {
-      entry = await promiseTimeout(this.registry.getEntry(publicKey, dataKey), 5000);
+      entry = await promiseTimeout(this.registry.getEntry(publicKey, dataKey), 30_000);
```
Comment (interstella-5555) [inline src/skydb.ts]: it will not stop the request so we might end up with hundreds of pending requests from one app, the timeout should not be in the client

if this is not possible right now, we should use axios `timeout` option, I **think** it will abort the request on timeout but I am not sure

File: src/skydb.ts [hunk 44]
```diff
@@ -2,15 +2,28 @@ import { pki } from "node-forge";
...
 import { RegistryEntry, SignedRegistryEntry } from "./registry";
-import { parseSkylink, promiseTimeout, trimUriPrefix, uriSkynetPrefix } from "./utils";
+import { parseSkylink, trimUriPrefix, uriSkynetPrefix } from "./utils";
 
+/**
+ * Gets the JSON object corresponding to the publicKey and dataKey.
+ * @param publicKey - The user public key.
+ * @param dataKey - The key of the data to fetch for the given user.
+ * @param [customOptions={}] - Additional settings that can optionally be set.
+ * @param [customOptions.timeout_seconds=30] - Timeout in seconds for the registry lookup.
```
Comment (interstella-5555) [inline src/skydb.ts]: `timeout_seconds` should be `timeout` and accept milliseconds - it's the standard in javascript to use milliseconds (check setTimeout as an example)

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

**Review by interstella-5555 — APPROVED** [review]
